extension/observer/dockerobserver/extension.go

```diff
@@ -166,92 +189,61 @@ func (d *dockerObserver) watchContainerEvents(listener observer.Notify, cacheRef
-		if endpoints, ok := d.existingEndpoints[cid]; ok {
-			for _, e := range endpoints {
-				existingEndpointsMap[e.ID] = e
-			}
-		}
-
-		// If the endpoint is present in existingEndpoints but is not
-		// present in latestEndpointsMap, then it needs to be removed.
-		for id, e := range existingEndpointsMap {
-			if _, ok := latestEndpointsMap[id]; !ok {
-				removedEndpoints = append(removedEndpoints, e)
-			}
-		}
-
-		// if the endpoint is present in latestEndpointsMap, check if it exists
-		// already in existingEndpoints.
-		for _, e := range latestEndpointsMap {
-			// If it does not exist already, it is a new endpoint. Add it.
-			if existingEndpoint, ok := existingEndpointsMap[e.ID]; !ok {
-				addedEndpoints = append(addedEndpoints, e)
-			} else if !reflect.DeepEqual(existingEndpoint, e) {
-				// if it already exists, see if it's equal.
-				// if it's not equal, update it
-				// if its equal, no-op.
-				updatedEndpoints = append(updatedEndpoints, e)
-			}
-		}
+func (d *dockerObserver) updateEndpointsForContainerID(cid string, latestContainerEndpoints map[observer.EndpointID]observer.Endpoint) {
+	if len(latestContainerEndpoints) == 0 {
+		// if latestContainerEndpoints is empty, we are removing all existing endpoints for the container
+		d.endpoints.Delete(cid)
+		return
+	}
 
-		// reset endpoints for this container
-		d.existingEndpoints[cid] = nil
-		// set the current known endpoints to the latest endpoints
-		for _, e := range latestEndpointsMap {
-			d.existingEndpoints[cid] = append(d.existingEndpoints[cid], e)
+	// map of EndpointID to endpoint to help with lookups
+	var existingEndpointsMap map[observer.EndpointID]observer.Endpoint
+	if endpointsMap, ok := d.endpoints.Load(cid); ok {
+		if existingEndpointsMap, ok = endpointsMap.(map[observer.EndpointID]observer.Endpoint); !ok {
+			d.logger.Info("unable to modify existing endpoints map", zap.Any("endpointsMap", endpointsMap))
+			return
 		}
 	} else {
-		// if latestEndpointsMap is nil, we are removing all endpoints for the container
-		removedEndpoints = append(removedEndpoints, d.existingEndpoints[cid]...)
-		delete(d.existingEndpoints, cid)
+		existingEndpointsMap = map[observer.EndpointID]observer.Endpoint{}
 	}
 
-	if len(removedEndpoints) > 0 {
-		listener.OnRemove(removedEndpoints)
+	// If the endpoint is present in existingEndpoints but is not
+	// present in latestContainerEndpoints, then it needs to be removed.
+	for id := range existingEndpointsMap {
+		if _, ok := latestContainerEndpoints[id]; !ok {
+			delete(existingEndpointsMap, id)
+		}
 	}
 
-	if len(addedEndpoints) > 0 {
-		listener.OnAdd(addedEndpoints)
+	// if the endpoint is present in containerEndpoints, check if it exists
+	// already in existingEndpoints.
+	for _, e := range latestContainerEndpoints {
+		existingEndpointsMap[e.ID] = e
 	}
 
-	if len(updatedEndpoints) > 0 {
-		listener.OnChange(updatedEndpoints)
-	}
+	d.endpoints.Store(cid, existingEndpointsMap)
```

Comment:
Isn't  `existingEndpointsMap` essentially a copy of `latestContainerEndpoints`? Can't we just do
```
d.endpoints.Store(cid, latestContainerEndpoints)
```
instead?

Also it doesn't look like the `latestContainerEndpoints` is used anywhere else apart from passing it to the `updateEndpointsForContainerID` call. Maybe we should call `containerEndpoints` right from `updateEndpointsForContainerID` function instead of passing `latestContainerEndpoints`?

Reply:
You're right and we no longer need to evaluate previous state since the endpoint watcher does it for us. After removing this it became clearer that this logic mirrors what's already available in the shared helper so most of it can be removed. Updated with this in mind.